manticore/core/cpu/binja.py

```diff
@@ -0,0 +1,1184 @@
+
+def _overflow_flag(res, right, left, size):
+    mask = 1 << (size - 1)
+    sign_r = (right & mask) == mask
+    sign_l = (left & mask) == mask
+    sign_res = (res & mask) == mask
+    return Operators.AND(sign_r ^ sign_l, sign_r ^ sign_res)
+
+def binja_platform_regmap(binja_arch_regs, binja_arch_flags, binja_arch_aliases,
+                          x86_cpu_regs):
+    all_x86_regs = [r.lower() for r in x86_cpu_regs]
+
+    # get all registers that are present already in the arch
+    x86_to_binja = {r.upper() : r for r in all_x86_regs
+                    if r in binja_arch_regs}
+
+    # map flag registers
+    for f in binja_arch_flags:
+        x86_to_binja[f.upper()] = f
+    x86_to_binja['IF'] = None
+
+    x86_to_binja['RSP'] = binja_arch_aliases['STACK']
+    x86_to_binja['RIP'] = binja_arch_aliases['PC']
+    x86_to_binja['GS'] = 'gsbase'
+    x86_to_binja['FS'] = 'fsbase'
+
+    for f in set(x86_cpu_regs) - set(x86_to_binja.keys()):
+        # no floating point or vector :/
+        if f.startswith("YMM") or f.startswith("FP"):
+            x86_to_binja[f.upper()] = None
+        elif f[-1] == 'H':
+            if f[:-1].lower() in binja_arch_regs:
+                x86_to_binja[f] = (f[:-1].lower(), 'high')
+            elif "r" + f[:-1].lower() in binja_arch_regs:
+                x86_to_binja[f] = ("r" + f[:-1].lower(), 'high')
+        elif f.upper() in binja_arch_aliases.keys():
+            x86_to_binja[f.upper()] = f.upper()
+
+    # still unmapped
+    # x86_to_binja ['EIP', 'IP', 'FRAME', 'EFLAGS', 'RFLAGS', 'TOP']
+    # binja_to_x86: ['fs', 'gs', 'mmXX', 'stXX']
+    #  print set(x86_cpu.all_registers) - set(x86_to_binja.keys())
+    binja_to_x86 = dict()
+    binja_to_x86["high_mappings"] = []
+    for x86_reg, binja_reg in x86_to_binja.items():
+        if isinstance(binja_reg, tuple):
+            binja_to_x86["high_mappings"].append(binja_reg[0])
+        elif isinstance(binja_reg, str):
+            binja_to_x86[binja_reg] = x86_reg
+
+    return x86_to_binja, binja_to_x86
+
+class BinjaCpu(Cpu):
+    '''
+    A Virtual CPU model for Binary Ninja's IL
+    '''
+
+    # Will be initialized based on the underlying Binja Architecture
+    max_instr_width = None
+    address_bit_size = None
+    machine = None
+    arch = None
+    mode = None
+
+    platform_cpu = None
+
+    def __init__(self, view, memory):
+        '''
+        Builds a CPU model.
+        :param arch: BinaryNinja arch.
+        '''
+        # get a platform specific CPU
+        platform_cpu = CpuFactory.get_cpu(memory, 'amd64')
+        # and mark it as virtual so as to not increment the PC in the
+        # @instruction decorator
+        platform_cpu.real_cpu = False
+        self.__class__.platform_cpu = platform_cpu
+        platform_regs = platform_cpu.all_registers
+        from binaryninja import Architecture
+        self.__class__.max_instr_width = view.arch.max_instr_length
```

Comment:
i just realized, shouldn't all these attributes which we currently have on the class, actually just be regular instance variables? these are attributes which should not be shared by all instances of BinjaCpu